proof/refine/ARM/Tcb_R.thy

```diff
@@ -794,49 +796,541 @@ lemma threadSet_valid_objs_tcbPriority_update:
-   apply (force simp: tcb_at_st_tcb_at[symmetric] state_relation_def
-                dest: pspace_relation_tcb_at intro: st_tcb_at_opeqI)
-  apply (force simp: state_relation_def elim: valid_objs'_maxDomain valid_objs'_maxPriority)
-  done *)
+lemma tcbEPDequeue_valid_SendEP:
+  "\<lbrace>valid_ep' (SendEP q) and K (t \<in> set q)\<rbrace> tcbEPDequeue t q \<lbrace>\<lambda>q'. valid_ep' (SendEP (t#q'))\<rbrace>"
+  apply (wpsimp simp: tcbEPDequeue_def valid_ep'_def)
+  apply (fastforce simp: findIndex_def findIndex'_app
+                   dest: in_set_takeD in_set_dropD findIndex_member)
+  done
+
+lemma tcbEPDequeue_valid_RecvEP:
+  "\<lbrace>valid_ep' (RecvEP q) and K (t \<in> set q)\<rbrace> tcbEPDequeue t q \<lbrace>\<lambda>q'. valid_ep' (RecvEP (t#q'))\<rbrace>"
+  apply (wpsimp simp: tcbEPDequeue_def valid_ep'_def)
+  apply (fastforce simp: findIndex_def findIndex'_app
+                   dest: in_set_takeD in_set_dropD findIndex_member)
+  done
+
+lemma tcbEPDequeue_valid_ep':
+  "\<lbrace>valid_ep' (updateEpQueue ep q) and K (ep \<noteq> IdleEP \<and> t \<in> set q)\<rbrace>
+   tcbEPDequeue t q
+   \<lbrace>\<lambda>q'. valid_ep' (updateEpQueue ep (t#q'))\<rbrace>"
+  by (cases ep) (wpsimp wp: tcbEPDequeue_valid_SendEP tcbEPDequeue_valid_RecvEP simp: updateEpQueue_def)+
+
+lemma tcbEPDequeueAppend_valid_ntfn'_rv:
+  "\<lbrace>valid_ntfn' ntfn and K (ntfnObj ntfn = WaitingNtfn qs \<and> t \<in> set qs)\<rbrace>
+   do qs' \<leftarrow> tcbEPDequeue t qs;
+      tcbEPAppend t qs'
+   od
+   \<lbrace>\<lambda>rv s. valid_ntfn' (ntfnObj_update (\<lambda>_. WaitingNtfn rv) ntfn) s\<rbrace>"
+  apply (simp only: tcbEPAppend_def tcbEPDequeue_def)
+  apply (wp tcbEPFindIndex_wp)
+  apply (rule conjI)
+   apply (clarsimp simp: valid_ntfn'_def split: option.split)
+  apply (clarsimp simp: valid_ntfn'_def simp del: imp_disjL dest!: findIndex_member)
+  apply (intro conjI; clarsimp?)
+          apply (fastforce dest: in_set_takeD in_set_dropD)
+         apply (fastforce dest: in_set_dropD)
+        apply (fastforce dest: in_set_dropD)
+       apply (fastforce dest: in_set_dropD)
+      apply (fastforce dest: in_set_takeD)
+     apply (clarsimp simp: Int_Un_distrib set_take_disj_set_drop_if_distinct)
+     apply (rule disjoint_subset_both[OF set_take_subset set_drop_subset])
+     apply (simp add: Int_commute)
+    apply (fastforce dest: in_set_takeD)
+   apply (clarsimp simp: Int_Un_distrib set_take_disj_set_drop_if_distinct)
+   apply (fastforce dest: in_set_takeD in_set_dropD)
+  apply (clarsimp split: option.split)
+  apply (case_tac ys; clarsimp)
+  done
+
+lemma reorderNtfn_invs':
+  "\<lbrace>invs' and st_tcb_at' (\<lambda>st. ntfnBlocked st = Some ntfnPtr) tptr\<rbrace>
+   reorderNtfn ntfnPtr tptr
+   \<lbrace>\<lambda>rv. invs'\<rbrace>"
+  apply (simp only: reorderNtfn_def)
+  apply (subst bind_assoc[symmetric, where m="tcbEPDequeue tptr _"])
+  apply (rule hoare_seq_ext | simp only: K_bind_def)+
```

Comment:
I think I've seen you use this `(rule hoare_seq_ext | simp something)+` pattern before. I'm curious: is this much different from just saying `wpsimp` to begin with? Do you see much benefit using this?

Reply:
`tcbEPDequeueAppend_valid_ntfn'_rv` applies over two statements. If I were to use `wp`, the instance of `bind_assoc` would be undone since the splitting takes priority over rule applications.